torch_nets/models.py

```diff
@@ -31,59 +45,116 @@ class ANN(Module):  # pylint: disable=too-many-instance-attributes
 
     """
 
-    def __init__(
+    def __init__(  # pylint: disable=too-many-arguments,too-many-locals
         self,
         n_in: int = 61,
         n_out: int = 148,
-        neurons=128,
-        dropout=0.0,
+        n_layers: int = 5,
+        neurons: int = 128,
+        dropout: float = 0.0,
+        device: str = "cpu",
+        features_mean: Optional[Tensor] = None,
+        features_std: Optional[Tensor] = None,
+        outputs_mean: Optional[Tensor] = None,
+        outputs_std: Optional[Tensor] = None,
+        output_groups: Optional[Tensor] = None,
     ):
-        """Build ``ANN``."""
-        super().__init__()
-        self.linear1 = Linear(n_in, neurons)
-        self.linear2 = Linear(neurons, neurons)
-        self.linear3 = Linear(neurons, neurons)
-        self.linear4 = Linear(neurons, neurons)
-        self.linear5 = Linear(neurons, n_out)
-
-        self.lin_drop1 = Dropout(dropout)
-        self.lin_drop2 = Dropout(dropout)
-        self.lin_drop3 = Dropout(dropout)
-        self.lin_drop4 = Dropout(dropout)
-
-    def forward(self, batch):
-        """Pass ``batch`` through the model.
+        """Initialize the ANN model."""
+        dims = [n_in] + [neurons] * (n_layers - 1) + [n_out]
+        layers = []
+
+        for i in range(n_layers):
+            layers.append(nn.Linear(dims[i], dims[i + 1]))
+            if i < n_layers - 1:
+                layers.append(nn.ReLU())  # type: ignore
+                layers.append(nn.Dropout(dropout))  # type: ignore
+
+        super().__init__(*layers)
+
+        fmean = fstd = omean = ostd = None
+
+        if features_mean is not None:
+            assert features_std is not None
+            assert len(features_mean) == len(features_std)
+            fmean = torch.tensor(features_mean)
+            fstd = torch.tensor(features_std)
+
+        if outputs_mean is not None:
+            assert outputs_std is not None
+            assert len(outputs_mean) == len(outputs_std)
+            if output_groups is None:
+                omean = torch.tensor(outputs_mean)
+                ostd = torch.tensor(outputs_std)
+            else:
+                assert len(output_groups) == len(outputs_mean)
+                omean = torch.tensor(
+                    [x for x, g in zip(outputs_mean, output_groups) for _ in range(g)]
+                )
+                ostd = torch.tensor(
+                    [x for x, g in zip(outputs_std, output_groups) for _ in range(g)]
+                )
+
+        self.register_buffer("features_mean", fmean)
+        self.register_buffer("features_std", fstd)
+        self.register_buffer("outputs_mean", omean)
+        self.register_buffer("outputs_std", ostd)
+
+        self.to(torch.device(device))
+
+    def forward(self, input: Tensor):  # pylint: disable=redefined-builtin
+        """Pass the input through the model.
```

Comment:
Perhaps extend here to say that we are passing through the layers, but overriding `forward` to also apply scaling of input and output data?

Reply:
I've extended the docstring

Reply:
Much better and understandable, thank you.